### ISSUE TITLE
Crash in lfs after module is unloaded

### ISSUE THREAD

**Issue by vhenriet-sfy** [LuaJIT/LuaJIT#961]
I have a crash with the snippet:

```lua
local lfs = require("lfs")

local foo = debug.setmetatable(newproxy(), {
    __gc = function()
        lfs.dir(".")
    end
})
```

I've tried with branch 2.0 and 2.1 with the same result.

1. lfs is creating a metatable in the registry: `luaL_newmetatable(L, DIR_METATABLE);` and set various stuff in it.
2. main program end, `foo` is garbage collected, __gc metamethod is called.
3. `lfs.dir()` is called which create a new userdata, get the DIR_METATABLE table and affect it to the new uyserdate.

```c
d = (dir_data *) lua_newuserdata(L, sizeof(dir_data));
  luaL_getmetatable(L, DIR_METATABLE);
  lua_setmetatable(L, -2);
```

Then a segfault happen.

If I recreate the metatable before getting it => no segfault
If I force the garbage collection before the end => no segfault

I guess that luajit has already destroy the DIR_METATABLE table which make the crash. But if I call `lua_istable(L, -1)`, it's true so I can't check that the table does'nt exist anymore.

**Comment by MikePall** [comment]
This code creates a new userdata (the dir iterator object) in the final `__gc` metamethod. The dir iterator itself has a `__gc` metamethod. However the iterator doesn't anchor the lfs module.
So it can happen that the module is unloaded before the dir iterator `___gc` metamethod is called. And that C function still points into the (now unmapped) memory of the shared library of the module, which of course crashes.

In other words: don't do that. It's generally not a good idea to create new objects with finalizers in the final `__gc` metamethod. Because most modules are not prepared for that.
